### ISSUE TITLE
Instability in BL with Periodic Boundary Conditions

### ISSUE THREAD

**Issue by jf-thomas** [su2code/SU2#1467]
**Describe the bug**
In a RANS computation of a 2.5D section of a wing, the periodic plane exhibits oscillations of pressure in the boundary layer. The issue is only present on the periodic face, not in the volume. Using a CFL of 1 helps but the instability does not disappear. A CFL of 10 leads to a crash of the computation with multiple spots like on the screenshot. Would it be linked to issue #1090  ?

![pres_per_face](https://user-images.githubusercontent.com/91184648/145236024-fc713cbd-66d8-40b3-9036-7c1925f10229.png)
![pres_cut](https://user-images.githubusercontent.com/91184648/145236048-a91ca27a-c1c9-4d6c-9639-8b34ea5de307.png)

Here is the config file:
[hl_3d_landing.txt](https://github.com/su2code/SU2/files/7677092/hl_3d_landing.txt)

**Hardware and software:**
 - OS:  Linux Red Hat ELS 6.5
 - Nodes: Intel Xeon Ivy bridge
 - C++ compiler and version: intel/2018.0.1
 - MPI implementation and version: intelmpi/2018.0.1
 - SU2 Version: v7.2.1

Thank you for looking at this issue and for any help.

**Comment by jf-thomas** [comment]
I made a light test case to reproduce the issue. It's a cylinder in a flowfield. It's only 112k cells.

![mesh](https://user-images.githubusercontent.com/91184648/145433376-ec7d4309-acc6-43d2-a172-39502e823147.png)
![swpeptCylinder](https://user-images.githubusercontent.com/91184648/145433398-97123613-824f-4f81-a143-aaae790ce252.png)
![residual_energy](https://user-images.githubusercontent.com/91184648/145433426-e6fc38f8-8482-47be-bd52-dca747b3e3e4.png)

The residuals never converge:
![residual_convergence](https://user-images.githubusercontent.com/91184648/145432900-0c2d7432-807b-4fe4-b24d-abf01cdc7bc4.png)
Here is the archive containing the case:
[cyl.zip](https://github.com/su2code/SU2/files/7686715/cyl.zip)

**Comment by pcarruscag** [comment]
Hi,
Can you try the develop branch or SA instead of SA_NEG. We fixed an issue in SA_NEG that could cause this type of oscillation.
However, I suspect it is due to issues found in #763 and that @TobiKattmann has also run into when implementing periodic BC's for heat-transfer problems.

How do you feel about doing some c++ hacking? We have an idea of how to improve things, just not enough bandwidth to do it.

**Comment by jf-thomas** [comment]
Hi,
I tried with SA instead of SA_NEG, I get the same behavior. I understand that I don't need to test the develop branch then.

I am ok with some c++ dev but I am not an expert. How would you improve the current implementation?

Thanks,
Jean-François

**Comment by pcarruscag** [comment]
Hi Jean,
Sorry for the delay. The hypothesis I have for the less robust behavior of periodic BC's is that the linear solver is not completely aware of the periodicity. In fact after the linear solve we have to force the matching nodes to have the same value.
This is done in CFVMFlowSolverBase.hpp::CompleteImplicitIteration_impl, in the call to InitiatePeriodicComms.
The idea is to make the linear solver aware of the periodicity, to do so would require including periodic communications in the preconditioners and the matrix-vector product.
These are all in CSysMatrix.cpp, so before each of the `/*--- MPI Parallelization ---*/` bits we would need periodic comms, for preconditioners these comms would simply make the values equal, like in CSolver::InitiatePeriodicComms(PERIODIC_IMPLICIT) whereas for the matrix-vector product we need to add the values (effectively periodicity splits a row of the matrix into two "half rows") and this would be similar to what is done with the linear residual in CSolver::InitiatePeriodicComms(PERIODIC_RESIDUAL).
I think @TobiKattmann was also interested in having a look into this.

**Comment by TobiKattmann** [comment]
Yes, indeed. I'll open a (rather empty) Draft PR soon in order to pressure myself a bit to get this test going. I am sorry for the delay :(

**Comment by jf-thomas** [comment]
Hi @pcarruscag, thank you for the development. I am testing it right now, I already noticed the struggle to converge the linear system. I lowered the CFL and relaxed the tolerance to 0.1.

**Comment by pcarruscag** [comment]
Hi @jf-thomas, was there a benefit in the end? Were you able to get better convergence in the cases you were working on?

**Comment by jf-thomas** [comment]
Hi @pcarruscag, sorry for the delay, I had a week off. I can confirm that it converges now. The only drawback is the longer convergence of the linear system. Thank you.

**Comment by pcarruscag** [comment]
Hi @jf-thomas no worries, thank you for the update. Just to clarify, are you referring just to the small example problem? Or also to the 2.5D wing section?

**Comment by jf-thomas** [comment]
I am referring to the 2.5D wing section.

**Comment by pcarruscag** [comment]
Brilliant, thanks. I'll put some time into finishing that pull request.

**Comment by TobiKattmann** [comment]
Still relevant. A fix is proposed in #1585, but due to problems with the discrete adjoint that is not merged yet. Works very fine for primal problems though.

**Comment by jf-thomas** [comment]
Hi all, with the release of v8.3.0, I've seen that a fix for periodic condition has been included (#2520). I did a run on the cylinder test case. It improves the convergence and reduces by two orders of magnitude the intensity of the residual oscillations. But it might also come from the jump from v7.2 to v8.3.

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/7ea466a8-961e-469a-a35c-b264b604a65c" />

<img width="1560" height="811" alt="Image" src="https://github.com/user-attachments/assets/853fab05-f1ae-4573-9239-d1a2925b446c" />
